### PR TITLE
Medical Engine - Improve damage calculation for explosive-resistant armor

### LINKED ISSUE #7799: Explosive Armor Value Causing Bullet Sponges
**Mods (complete and add to the following information):**
- **Arma 3:** ` 1.98 Stable` 
- **CBA:** `3.15 Stable` 
- **ACE3:** `3.13.3 Stable` 
_Make sure to reproduce the issue with only CBA and ACE3 on a newly created mission!_

**Description:**
EOD/Explosive/GL Vests Spike Armor Values

**Steps to reproduce:**
Place down 1 Standard Rifleman and 1 Grenadier/Team Lead for AAF
Place down Player Character with a 5.56 Weapon
Shoot Rifleman then shoot Grenadier in the Chest until they die

**Expected behavior:**
Grenadier takes way longer then they should too die

**Where did the issue occur?**
- Editor/Zeus Multiplayer Editor/Zeus Singleplayer

**Log Files:**
The issue didn't cause me to crash

**Additional context:**
Note that GA Carrier Lite vest has an armor value of 18 whereas the GA carrier GL rigs have an armor value of 80 which Is probably the Issue

**Screenshots:**
I have two Videos of it https://youtu.be/RiFA4CCaGcE and https://t.co/v3DPQ0V5Ay?amp=1

### LINKED ISSUE #8008: The Carrier Special Rig and Carrier GL Rig seemingly have swapped ballistic defense values
**Mods (complete and add to the following information):**
- **Arma 3:** `2.00`
- **CBA:** `3.15.2`
- **ACE3:** `3.13.5`

**Description:**
In spite of the in-game ballistic defense rating (not explosive), the Special rig does not stop bullets as efficiently (sometimes not at all) as the GL rig, even though the latter has lower ballistic defense.

**Steps to reproduce:**
1. Set up AI victims with both the Special rig and GL rig to shoot at.
2. Shoot AI victims until they die.

**Expected behavior:**
- I expected the AI victims with the Special rig to survive bullets for longer than those with the GL rig

**Where did the issue occur?**
- Editor (Singleplayer) as test environment

**Log Files:**
- [https://pastebin.com/aGmb4LSp](url)

**Additional context:**
- I'm running the medical setting "Only large hits to vital organs" and Critical Damage Threshold (both AI and Player) of 0.00

**Screenshots:**
<details>
<summary>GL rig screenshot</summary>

</details>
<details>
<summary>Special rig screenshot</summary>

</details>

### LINKED ISSUE #8315: No damage to highly armored targets
**Mods (complete and add to the following information):**
- **Arma 3:** `stable` 
- **CBA:** `3.15.4` 
- **ACE3:** `3.13.6` 

**Description:**
It takes a ridiculous amount of bullets to kill an AI with a high armor vest like a 'GA Carrier GL Rig' or a 'Carrier Special Rig', even surviving more than 4 bullets from a M200 Intervention. 5.45 can't even penetrate regular Plate Carriers.

**Steps to reproduce:**

1. Place an AI and equip it with a high armor vest.
2. Try to kill it with an APEX 416 or AKS-74U

**Expected behavior:**
AI should probably go down within 4-5 or even 6 bullets? 10? 

**Actual behavior:**
You can dump the whole mag and the AI doesn't even care, even when using 7.62 or .308.

**Where did the issue occur?**
- Editor (Singleplayer) / Virtual Arsenal

**Log Files:**
- https://pastebin.com/E72UZ0bF

### LINKED ISSUE #8956: Bullet did not do the damage to some vest
Mods 

ARMA 3: Current Live Build as of 5/24/2022
ACE3: Version 3.15.0
CBA:v3.15.7.220517
Description:
When I do some shooting in the ACE Arsenal ，I find some small arms can not do the damage to some vest .I shoot the  3 mags of ammo to some vest but in medical menu it only turn blue .You can not heal the body.
This is an image
Small Velocity Wound for 轻度挫伤
Where did the issue occur?

 Self-Hosted Multiplayer / Singleplayer / Editor (Singleplayer) / Editor (Multiplayer) / Virtual Arsenal
Log Files:
~~~~~~~~~ACE Debug~~~~~~~~~
[Time: 47.4] [CBA_missionTime: 47.2] [tickTime: 89.1]
------Performance------
diag_fps = 58.3942
count cba_common_waitAndExecArray = 4
count cba_common_waitUntilAndExecArray = 0
count cba_common_perFrameHandlerArray = 14 (max 128)
count diag_activeSQFScripts = 3
count diag_activeSQSScripts = 0
count diag_activeMissionFSMs = 1

------Player------
typeOf = C_man_1
animationState = amovpercmstpsraswrfldnon

------ACE's CBA Settings [569 Total] [Only Non-Defaults]------
ace_advanced_ballistics_enabled - true
ace_nightvision_aimDownSightsBlur - 0.284092
ace_nightvision_noiseScaling - 0.266685
ace_overheating_enabled - false
acex_headless_enabled - true

------Array Info (count >= 100)------
ace_medical_blood_blooddrops - ARRAY SIZE: 128
ace_common_hexarray - ARRAY SIZE: 256
cba_wtab - ARRAY SIZE: 128
ace_medical_gui_selfinteractionactions - ARRAY SIZE: 128
cba_ktab - ARRAY SIZE: 128
cba_common_pfhhandles - ARRAY SIZE: 128
cba_events_keyupstates - ARRAY SIZE: 270
ace_common_clearcache_ace_interact_menu_clearconditioncaches - ARRAY SIZE: 216
cba_common_addons - ARRAY SIZE: 1669
cba_settings_allsettings - ARRAY SIZE: 578
cba_events_keydownstates - ARRAY SIZE: 270
cba_xeh_alleventhandlers - ARRAY SIZE: 347
ace_casings_casings - ARRAY SIZE: 108
ace_interact_menu_inheritedactionsman - ARRAY SIZE: 256
cba_settings_needrestart - ARRAY SIZE: 168
cba_ytab - ARRAY SIZE: 128

### LINKED ISSUE #9044: Ace causes abnormal armor bulletproof of the RHS module
**Mods (complete and add to the following information):**
- **Arma 3:** `2.10`
- **CBA:** `v3.15.8.220912` 
- **ACE3:** `3.15.1` 
- **RHSUSAF**`0.5.6`
- **RHSGREF**`0.5.6`
- **RHSAFRF**`0.5.6`
- 
**Description:**
When using ACE and RHS, the bullet rating of the RHS armor will be changed directly to the highest bullet rating. The armor cannot be penetrated with any 7.62 or 5.56 caliber round. No matter how many bullets are fired at the AI chest, it will not deal any damage other than contusion.

**Steps to reproduce:**
Load the above module and enter the game to test

**Expected behavior:**
Bulletproof data is realistic(like normal)

**Where did the issue occur?**
anywhere

**Video:**
https://youtu.be/hEypVxJ1vDY

### LINKED ISSUE #9063: Carrier GL Rig & GA Carrier GL Rig Bullet Immunity
**Mods (complete and add to the following information):**
- **Arma 3:** `2.10 Stable` [e.g. 1.00 stable, rc,  dev]
- **CBA:** `3.15.8 Stable` [e.g. 3.0.0 stable, commit hash]
- **ACE3:** `3.15.2 Stable` [eg. 3.0.0 stable, commit hash]

**Description:**
Targets wearing either Carrier GL Rig or GA Carrier GL Rig can live through multiple mags worth of bullets when hit center mass.

**Steps to reproduce:**
_Follow [https://ace3.acemod.org/img/wiki/user/issue_flowchart.webp](this flowchart)!_

1. Create a new mission. For this, I used Eden Editor on Altis
2. Place two AI targets that you intend to shoot. Equip them with Carrier GL Rig or GA Carrier GL Rig
3. Get a gun. _For my testing, I used an MX with 6.5mm_
4. Shoot the target center mass. _For my testing, I proned 30m away from the target, used bipod, and unloaded mag after mag into the chest with no result other than the target having bruises_

**Expected behavior:**
If the Carrier Special Rig, which has full ballistic protection, can soak 10-15 bullets before the target dies, then the Carrier GL Rig & GA Carrier GL Rig shouldn't soak more.

**Where did the issue occur?**
- Replicated in BI VA, ACE VA, Eden Editor SP and Eden Editor Self-Hosted MP

**Log Files:**
https://gist.github.com/JFCO-Highlander/c59e9e7070acae49dca023aeb1fffe8d

**Additional context:**

**Screenshots:**
https://www.youtube.com/watch?v=Q4MCsfYcs5Y

### PR DESCRIPTION
**Pull request by LinkIsGrim**
**When merged this pull request will:**
- Scales high armor values using passthrough to prevent gear with high explosive resistance (armor) becoming immune to bullets. See #9063 and others.
- Close #9063.
- Close #9044.
- Close #8315.
- Close #8008.
- Close #7799.
- Close #8956.

~~I've only done limited testing but this doesn't seem to break fire/fall damage/drowning/explosives or have anything beyond the intended effect on wound creation.~~

Doesn't break anything.

This also improves protection from vests and helmets against pistol calibers, especially the lower tier tactical vests and armored uniforms as a side effect, but this is closer to expected performance when using vanilla as a reference.

image

### IMPORTANT

- [ ] If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [x] [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- [x] Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

**Comment by Salluci** [comment]
Here's some math: 
image

**Comment by Salluci** [comment]
Marked as WIP. High damage diverts to limbs again.

File: addons/medical_engine/functions/fnc_getItemArmor.sqf [hunk 1]
```diff
@@ -38,14 +39,22 @@ if (isNil "_armor") then {
...
         private _entry = configProperties [_itemInfo >> "HitpointsProtectionInfo", _condition] param [0, configNull];
-
-        _armor = getNumber (_entry >> "armor");
+        if (!isNull _entry) then {
+            _armor = getNumber (_entry >> "armor");
+            _passThrough = (0.01 max getNumber (_entry >> "passThrough"));
+        };
     };
 
+    // Scale armor using passthrough to fix explosive resistant armor (#9063)
```
Comment (rautamiekka) [inline addons/medical_engine/functions/fnc_getItemArmor.sqf:53]:
```suggestion
    // Scale armor using passthrough to fix explosive-resistant armor (#9063)
```

File: addons/medical_engine/functions/fnc_getItemArmor.sqf [hunk 2]
```diff
@@ -38,14 +39,22 @@ if (isNil "_armor") then {
...
+    // Skip scaling for items that don't cover the hitpoint to prevent infinite armor
+    if (_armor != 0) then {
```
Comment (rautamiekka) [inline addons/medical_engine/functions/fnc_getItemArmor.sqf:55]:
```suggestion
    if (_armor isNotEqualTo 0) then {
```

File: addons/medical_engine/functions/fnc_getItemArmor.sqf [hunk 3]
```diff
@@ -38,14 +39,22 @@ if (isNil "_armor") then {
...
+            _passThrough = (0.01 max getNumber (_entry >> "passThrough")); // prevent dividing by 0
         };
     } else {
         private _condition = format ["getText (_x >> 'hitpointName') == '%1'", _hitpoint];
         private _entry = configProperties [_itemInfo >> "HitpointsProtectionInfo", _condition] param [0, configNull];
-
-        _armor = getNumber (_entry >> "armor");
+        if (!isNull _entry) then {
+            _armor = getNumber (_entry >> "armor");
+            _passThrough = (0.01 max getNumber (_entry >> "passThrough"));
```
Comment (rautamiekka) [inline addons/medical_engine/functions/fnc_getItemArmor.sqf:49]:
```suggestion
            _passThrough = 0.01 max getNumber (_entry >> "passThrough");
```

**Comment by Salluci** [comment]
Fixed the sorting, though if anyone has a more elegant solution, I'll take it.

**Review by Mike-MF — APPROVED** [review]
I tested it, it works as advertised. Someone other than me should probably review the code though.

**Review by veteran29 — APPROVED** [review]

**Comment by LinkIsGrim** [comment]
FINALLY, BEGONE ISSUES

**Pull request merged**